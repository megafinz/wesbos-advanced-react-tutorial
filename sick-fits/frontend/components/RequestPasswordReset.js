import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';
import Form from './styles/Form';

export const REQUEST_PASSWORD_RESET_MUTATION = gql`
  mutation REQUEST_PASSWORD_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      message
    }
  }
`;

export default function RequestPasswordReset() {
  const { inputs, handleChange, clearForm } = useForm({
    email: '',
  });
  const [requestPasswordReset, { data, error, loading }] = useMutation(
    REQUEST_PASSWORD_RESET_MUTATION,
    {
      variables: inputs,
    }
  );
  return (
    <Form
      method="post"
      onSubmit={async (e) => {
        e.preventDefault();
        await requestPasswordReset();
        clearForm();
      }}
    >
      <h2>Request Password Reset</h2>
      <DisplayError error={error} />
      {data?.sendUserPasswordResetLink === null && (
        <p>Success! Check your email for a link.</p>
      )}
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your Email Address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
      </fieldset>
      <button type="submit">Request Reset!</button>
    </Form>
  );
}
