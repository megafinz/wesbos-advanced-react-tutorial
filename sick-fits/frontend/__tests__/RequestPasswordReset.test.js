import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestPasswordReset, {
  REQUEST_PASSWORD_RESET_MUTATION,
} from '../components/RequestPasswordReset';

const mocks = [
  {
    request: {
      query: REQUEST_PASSWORD_RESET_MUTATION,
      variables: {
        email: 'fake-user@example.com',
      },
    },
    result: {
      data: {
        sendUserPasswordResetLink: null,
      },
    },
  },
];

describe('<RequestPasswordReset />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestPasswordReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation when submitted', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <RequestPasswordReset />
      </MockedProvider>
    );
    userEvent.type(
      screen.getByPlaceholderText(/email/i),
      'fake-user@example.com'
    );
    userEvent.click(screen.getByText(/Request Reset/));
    await screen.findByText(/Success/);
  });
});
