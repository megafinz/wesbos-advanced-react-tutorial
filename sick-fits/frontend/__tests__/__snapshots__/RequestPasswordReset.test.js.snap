// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestPasswordReset /> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 JYcwP"
    method="post"
  >
    <h2>
      Request Password Reset
    </h2>
    <fieldset
      aria-busy="false"
    >
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
    </fieldset>
    <button
      type="submit"
    >
      Request Reset!
    </button>
  </form>
</div>
`;
